drivers/flash/flash_stm32h7x.c

```diff
@@ -31,11 +31,11 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
 #define STM32H7_FLASH_TIMEOUT	\
 	(2 * DT_PROP(DT_INST(0, st_stm32_nv_flash), max_erase_time))
 
-#ifdef CONFIG_CPU_CORTEX_M4
-#error Flash driver on M4 core is not supported yet
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && defined(CONFIG_CPU_CORTEX_M4)
+#define REAL_FLASH_SIZE_KB KB(2048)
```

Comment:
question: could this be retrieve from the DTS: the chosen <zephyr,flash> and its size which is `DT_SIZE_K(1024)`in the dts/arm/st/h7/stm32h747Xi_m4.dtsi

Reply:
@FRASTM REAL_FLASH_SIZE_KB  is the actual size of the Flash device which is 2048 and is read from the register, where as DT value is the bank size used by M4 which is different. The way it is used in the driver, it is the device flash size. So this looks fine to me.

Reply:
I think I need to change the if check as below so that for other SoCs, it works as before

#if defined(CONFIG_SOC_STM32H747XX) && defined(CONFIG_CPU_CORTEX_M4)
#define REAL_FLASH_SIZE_KB KB(2048)
#endif

Reply:
If this depends on SoC, it should be configurable somehow (in a way that doesn't need to be updated each time a new target is added), hence proposed code is not possible.

If it should be configured, device tree should be used as is the way in Zephyr to describe hardware.
Alternatively, if it provides a solution much more convenient than device tree, you can use input from HAL.
But Kconfig shouldn't be used for this purpose.